File: turn-orchestrator/src/persistence.rs
//! State load/save helpers. Each helper runs one or two `state::*` triggers
//! and never panics; missing keys deserialise to defaults so callers can
//! treat first-time and retry paths the same way.

use harness_types::{AgentMessage, FunctionCall, FunctionResult};
use iii_sdk::{TriggerRequest, Value, III};
use serde_json::{json, Value as JsonValue};

use crate::state::{
    cwd_index_key, cwd_key, function_schemas_key, messages_key, run_request_key, sandbox_id_key,
    tool_schemas_key, turn_state_key, TurnStateRecord,
};

const STATE_SCOPE: &str = "agent";

pub async fn load_record(iii: &III, session_id: &str) -> Option<TurnStateRecord> {
    let key = turn_state_key(session_id);
    let value = state_get(iii, &key).await?;
    serde_json::from_value(value).ok()
}

pub async fn save_record(iii: &III, record: &TurnStateRecord) {
    let key = turn_state_key(&record.session_id);
    if let Ok(value) = serde_json::to_value(record) {
        state_set(iii, &key, value).await;
    }
}

pub async fn load_messages(iii: &III, session_id: &str) -> Vec<AgentMessage> {
    let key = messages_key(session_id);
    let Some(value) = state_get(iii, &key).await else {
        return Vec::new();
    };
    serde_json::from_value(value).unwrap_or_default()
}

pub async fn save_messages(iii: &III, session_id: &str, messages: &[AgentMessage]) {
    let key = messages_key(session_id);
    if let Ok(value) = serde_json::to_value(messages) {
        state_set(iii, &key, value).await;
    }
    // Best-effort mirror to session-tree. Failure does not abort the turn.
    mirror_messages_to_session_tree(iii, session_id, messages).await;
}

async fn mirror_messages_to_session_tree(iii: &III, session_id: &str, messages: &[AgentMessage]) {
    let last_key = crate::state::last_session_tree_len_key(session_id);
    let already_mirrored = state_get(iii, &last_key)
        .await
        .and_then(|v| v.as_u64())
        .unwrap_or(0) as usize;

    if messages.len() <= already_mirrored {
        return;
    }

    if already_mirrored == 0 {
        if let Err(e) = iii
            .trigger(TriggerRequest {
                function_id: "session-tree::ensure".into(),
                payload: json!({ "session_id": session_id }),
                action: None,
                timeout_ms: None,
            })
            .await
        {
            tracing::warn!(
                error = %e,
                %session_id,
                "turn-orchestrator: session-tree::ensure failed; mirror skipped"
            );
            return;
        }
    }

    // Find the current leaf entry_id by reading session-tree state — needed
    // so subsequent appends thread parent_id correctly. For a fresh session
    // (already_mirrored == 0), parent_id starts as None.
    let mut last_appended: Option<String> = None;
    if already_mirrored > 0 {
        match iii
            .trigger(TriggerRequest {
                function_id: "session-tree::messages".into(),
                payload: json!({ "session_id": session_id }),
                action: None,
                timeout_ms: None,
            })
            .await
        {
            Ok(resp) => {
                last_appended = resp
                    .get("messages")
                    .and_then(|m| m.as_array())
                    .and_then(|arr| arr.last())
                    .and_then(|last| last.get("entry_id"))
                    .and_then(|v| v.as_str())
                    .map(String::from);
            }
            Err(e) => {
                tracing::warn!(
                    error = %e,
                    %session_id,
                    "turn-orchestrator: session-tree::messages read failed mid-mirror; skipping append batch to avoid orphaning"
                );
                return;
            }
        }
    }

    for msg in &messages[already_mirrored..] {
        let payload = json!({
            "session_id": session_id,
            "parent_id": last_appended,
            "message": msg,
        });
        match iii
            .trigger(TriggerRequest {
                function_id: "session-tree::append".into(),
                payload,
                action: None,
                timeout_ms: None,
            })
            .await
        {
            Ok(resp) => {
                last_appended = resp
                    .get("entry_id")
                    .and_then(|v| v.as_str())
                    .map(String::from);
            }
            Err(e) => {
                tracing::warn!(
                    error = %e,
                    %session_id,
                    "turn-orchestrator: session-tree::append mirror failed"
                );
                return;
            }
        }
    }

    state_set(iii, &last_key, json!(messages.len())).await;
}

pub async fn save_run_request(iii: &III, session_id: &str, request: JsonValue) {
    let key = run_request_key(session_id);
    state_set(iii, &key, request).await;
}

pub async fn load_run_request(iii: &III, session_id: &str) -> JsonValue {
    state_get(iii, &run_request_key(session_id))
        .await
        .unwrap_or_else(|| json!({}))
}

pub async fn save_cwd(iii: &III, session_id: &str, cwd: &str) {
    state_set(
        iii,
        &cwd_key(session_id),
        JsonValue::String(cwd.to_string()),
    )
    .await;
}

pub async fn load_cwd(iii: &III, session_id: &str) -> Option<String> {
    state_get(iii, &cwd_key(session_id))
        .await
        .and_then(|v| v.as_str().map(str::to_string))
}

pub async fn save_cwd_index(iii: &III, cwd_hash: &str, session_id: &str) {
    state_set(
        iii,
        &cwd_index_key(cwd_hash),
        JsonValue::String(session_id.to_string()),
    )
    .await;
}

pub async fn save_sandbox_id(iii: &III, session_id: &str, sandbox_id: Option<&str>) {
    let key = sandbox_id_key(session_id);
    let value = sandbox_id.map_or(JsonValue::Null, |s| JsonValue::String(s.to_string()));
    state_set(iii, &key, value).await;
}

pub async fn load_sandbox_id(iii: &III, session_id: &str) -> Option<String> {
    state_get(iii, &sandbox_id_key(session_id))
        .await
        .and_then(|v| v.as_str().map(str::to_string))
}

pub async fn save_function_schemas(iii: &III, session_id: &str, schemas: JsonValue) {
    state_set(iii, &function_schemas_key(session_id), schemas).await;
}

/// Load function catalog JSON; falls back to legacy `tool_schemas` key when the new key is absent.
pub async fn load_function_schemas(iii: &III, session_id: &str) -> JsonValue {
    let new_key = function_schemas_key(session_id);
    match state_get(iii, &new_key).await {
        Some(v) => v,
        None => state_get(iii, &tool_schemas_key(session_id))
            .await
            .unwrap_or_else(|| json!([])),
    }
}

/// Back-compat name for callers being migrated — prefer [`save_function_schemas`].
#[inline]
pub async fn save_tool_schemas(iii: &III, session_id: &str, schemas: JsonValue) {
    save_function_schemas(iii, session_id, schemas).await;
}

/// Back-compat — prefer [`load_function_schemas`].
#[inline]
pub async fn load_tool_schemas(iii: &III, session_id: &str) -> JsonValue {
    load_function_schemas(iii, session_id).await
}

async fn state_get(iii: &III, key: &str) -> Option<Value> {
    match iii
        .trigger(TriggerRequest {
            function_id: "state::get".into(),
            payload: json!({ "scope": STATE_SCOPE, "key": key }),
            action: None,
            timeout_ms: None,
        })
        .await
    {
        Ok(v) if v.is_null() => None,
        Ok(v) => Some(v),
        Err(e) => {
            tracing::warn!(error = %e, %key, "turn-orchestrator: state::get failed");
            None
        }
    }
}

async fn state_set(iii: &III, key: &str, value: Value) {
    if let Err(e) = iii
        .trigger(TriggerRequest {
            function_id: "state::set".into(),
            payload: json!({ "scope": STATE_SCOPE, "key": key, "value": value }),
            action: None,
            timeout_ms: None,
        })
        .await
    {
        tracing::warn!(error = %e, %key, "turn-orchestrator: state::set failed");
    }
}

const PREPARED_KEY: &str = "function_prepared";
const EXECUTED_KEY: &str = "function_executed";
const LEGACY_PREPARED_KEY: &str = "tool_prepared";
const LEGACY_EXECUTED_KEY: &str = "tool_executed";

fn staging_key(session_id: &str, suffix: &str) -> String {
    format!("session/{session_id}/{suffix}")
}

async fn staging_get_with_legacy(
    iii: &III,
    session_id: &str,
    new_suffix: &str,
    legacy_suffix: &str,
) -> JsonValue {
    let new_k = staging_key(session_id, new_suffix);
    match state_get(iii, &new_k).await {
        Some(v) => v,
        None => state_get(iii, &staging_key(session_id, legacy_suffix))
            .await
            .unwrap_or_else(|| json!([])),
    }
}

pub async fn save_prepared_calls(
    iii: &III,
    session_id: &str,
    prepared: &[(FunctionCall, Option<FunctionResult>)],
) {
    let payload = serde_json::to_value(
        prepared
            .iter()
            .map(|(tc, pre)| json!({ "function_call": tc, "blocked": pre }))
            .collect::<Vec<_>>(),
    )
    .unwrap_or_else(|_| json!([]));
    state_set(iii, &staging_key(session_id, PREPARED_KEY), payload).await;
}

pub async fn load_prepared_calls(
    iii: &III,
    session_id: &str,
) -> Vec<(FunctionCall, Option<FunctionResult>)> {
    let value = staging_get_with_legacy(iii, session_id, PREPARED_KEY, LEGACY_PREPARED_KEY).await;
    let Some(arr) = value.as_array() else {
        return Vec::new();
    };
    arr.iter()
        .filter_map(|entry| {
            let fc = entry
                .get("function_call")
                .or_else(|| entry.get("tool_call"))
                .and_then(|v| serde_json::from_value::<FunctionCall>(v.clone()).ok())?;
            let pre = entry
                .get("blocked")
                .and_then(|v| serde_json::from_value::<Option<FunctionResult>>(v.clone()).ok())
                .unwrap_or(None);
            Some((fc, pre))
        })
        .collect()
}

pub async fn save_executed_calls(
    iii: &III,
    session_id: &str,
    executed: &[(FunctionCall, FunctionResult, bool)],
) {
    let payload = serde_json::to_value(
        executed
            .iter()
            .map(|(tc, r, e)| json!({ "function_call": tc, "result": r, "is_error": e }))
            .collect::<Vec<_>>(),
    )
    .unwrap_or_else(|_| json!([]));
    state_set(iii, &staging_key(session_id, EXECUTED_KEY), payload).await;
}

pub async fn load_executed_calls(
    iii: &III,
    session_id: &str,
) -> Vec<(FunctionCall, FunctionResult, bool)> {
    let value = staging_get_with_legacy(iii, session_id, EXECUTED_KEY, LEGACY_EXECUTED_KEY).await;
    let Some(arr) = value.as_array() else {
        return Vec::new();
    };
    arr.iter()
        .filter_map(|entry| {
            let fc = entry
                .get("function_call")
                .or_else(|| entry.get("tool_call"))
                .and_then(|v| serde_json::from_value::<FunctionCall>(v.clone()).ok())?;
            let r = serde_json::from_value::<FunctionResult>(entry.get("result")?.clone()).ok()?;
            let e = entry
                .get("is_error")
                .and_then(Value::as_bool)
                .unwrap_or(false);
            Some((fc, r, e))
        })
        .collect()
}

pub fn find_executed_call<'a>(
    executed: &'a [(FunctionCall, FunctionResult, bool)],
    function_call_id: &str,
) -> Option<&'a (FunctionCall, FunctionResult, bool)> {
    executed.iter().find(|(fc, _, _)| fc.id == function_call_id)
}

pub fn upsert_executed_call(
    executed: &mut Vec<(FunctionCall, FunctionResult, bool)>,
    entry: (FunctionCall, FunctionResult, bool),
) {
    if let Some(existing) = executed.iter_mut().find(|(fc, _, _)| fc.id == entry.0.id) {
        *existing = entry;
    } else {
        executed.push(entry);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::TurnState;
    use harness_types::{ContentBlock, TextContent};

    fn fc(id: &str, function_id: &str) -> FunctionCall {
        FunctionCall {
            id: id.into(),
            function_id: function_id.into(),
            arguments: json!({ "id": id }),
        }
    }

    fn func_result(text: &str) -> FunctionResult {
        FunctionResult {
            content: vec![ContentBlock::Text(TextContent { text: text.into() })],
            details: json!({ "text": text }),
            terminate: false,
        }
    }

    #[test]
    fn record_round_trips_through_json() {
        let mut r = TurnStateRecord::new("s1", Some(8));
        r.transition_to(TurnState::AwaitingAssistant);
        let v = serde_json::to_value(&r).unwrap();
        let back: TurnStateRecord = serde_json::from_value(v).unwrap();
        assert_eq!(back.state, TurnState::AwaitingAssistant);
        assert_eq!(back.session_id, "s1");
        assert_eq!(back.max_turns, Some(8));
    }

    #[test]
    fn find_executed_call_matches_function_call_id() {
        let executed = vec![
            (fc("tc-1", "read"), func_result("one"), false),
            (fc("tc-2", "write"), func_result("two"), true),
        ];

        let found = find_executed_call(&executed, "tc-2").expect("expected tc-2");

        assert_eq!(found.0.id, "tc-2");
        assert_eq!(found.0.function_id, "write");
        assert!(found.2);
        assert!(find_executed_call(&executed, "missing").is_none());
    }

    #[test]
    fn upsert_executed_call_preserves_order_and_replaces_existing() {
        let mut executed = vec![
            (fc("tc-1", "read"), func_result("one"), false),
            (fc("tc-2", "write"), func_result("two"), true),
        ];

        upsert_executed_call(
            &mut executed,
            (fc("tc-2", "write"), func_result("replacement"), false),
        );
        upsert_executed_call(
            &mut executed,
            (fc("tc-3", "list"), func_result("three"), false),
        );

        assert_eq!(executed.len(), 3);
        assert_eq!(executed[0].0.id, "tc-1");
        assert_eq!(executed[1].0.id, "tc-2");
        assert_eq!(executed[2].0.id, "tc-3");
        assert!(!executed[1].2);
        assert!(matches!(
            executed[1].1.content.first(),
            Some(ContentBlock::Text(text)) if text.text == "replacement"
        ));
    }
}

File: harness-node/src/turn-orchestrator/persistence.ts
import type { ISdk } from '../runtime/iii.js';
import { logger } from '../runtime/otel.js';
import type { AgentMessage } from '../types/agent-message.js';
import type { FunctionCall, FunctionResult } from '../types/function.js';
import {
  type TurnStateRecord,
  cwdIndexKey,
  cwdKey,
  functionSchemasKey,
  lastSessionTreeLenKey,
  messagesKey,
  runRequestKey,
  sandboxIdKey,
  toolSchemasKey,
  turnStateKey,
} from './state.js';

const SCOPE = 'agent';

async function stateGet(iii: ISdk, key: string): Promise<unknown | null> {
  try {
    const v = await iii.trigger<unknown, unknown>({
      function_id: 'state::get',
      payload: { scope: SCOPE, key },
    });
    return v === null || v === undefined ? null : v;
  } catch (err) {
    logger.warn('persistence state::get failed', { key, err: String(err) });
    return null;
  }
}

async function stateSet(iii: ISdk, key: string, value: unknown): Promise<void> {
  try {
    await iii.trigger<unknown, unknown>({
      function_id: 'state::set',
      payload: { scope: SCOPE, key, value },
    });
  } catch (err) {
    logger.warn('persistence state::set failed', { key, err: String(err) });
  }
}

export async function loadRecord(iii: ISdk, session_id: string): Promise<TurnStateRecord | null> {
  const v = await stateGet(iii, turnStateKey(session_id));
  if (!v || typeof v !== 'object') return null;
  return v as TurnStateRecord;
}

export async function saveRecord(iii: ISdk, rec: TurnStateRecord): Promise<void> {
  await stateSet(iii, turnStateKey(rec.session_id), rec);
}

export async function loadMessages(iii: ISdk, session_id: string): Promise<AgentMessage[]> {
  const v = await stateGet(iii, messagesKey(session_id));
  return Array.isArray(v) ? (v as AgentMessage[]) : [];
}

export async function saveMessages(
  iii: ISdk,
  session_id: string,
  messages: AgentMessage[],
): Promise<void> {
  await stateSet(iii, messagesKey(session_id), messages);
  await mirrorMessagesToSessionTree(iii, session_id, messages);
}

async function mirrorMessagesToSessionTree(
  iii: ISdk,
  session_id: string,
  messages: AgentMessage[],
): Promise<void> {
  const lastKey = lastSessionTreeLenKey(session_id);
  const last = await stateGet(iii, lastKey);
  const alreadyMirrored = typeof last === 'number' ? last : 0;
  if (messages.length <= alreadyMirrored) return;
  if (alreadyMirrored === 0) {
    try {
      await iii.trigger<unknown, unknown>({
        function_id: 'session-tree::ensure',
        payload: { session_id },
      });
    } catch (err) {
      logger.warn('session-tree::ensure failed; mirror skipped', {
        session_id,
        err: String(err),
      });
      return;
    }
  }
  let lastAppended: string | null = null;
  if (alreadyMirrored > 0) {
    try {
      const resp = await iii.trigger<unknown, { messages?: Array<{ entry_id?: string }> }>({
        function_id: 'session-tree::messages',
        payload: { session_id },
      });
      const arr = resp?.messages;
      if (Array.isArray(arr) && arr.length > 0) {
        const tail = arr[arr.length - 1];
        lastAppended = tail?.entry_id ?? null;
      }
    } catch (err) {
      logger.warn('session-tree::messages read failed mid-mirror; skipping', {
        session_id,
        err: String(err),
      });
      return;
    }
  }
  for (const msg of messages.slice(alreadyMirrored)) {
    try {
      const resp = await iii.trigger<unknown, { entry_id?: string }>({
        function_id: 'session-tree::append',
        payload: { session_id, parent_id: lastAppended, message: msg },
      });
      lastAppended = resp?.entry_id ?? lastAppended;
    } catch (err) {
      logger.warn('session-tree::append mirror failed', { session_id, err: String(err) });
      return;
    }
  }
  await stateSet(iii, lastKey, messages.length);
}

export async function saveRunRequest(
  iii: ISdk,
  session_id: string,
  request: unknown,
): Promise<void> {
  await stateSet(iii, runRequestKey(session_id), request);
}

export async function loadRunRequest(
  iii: ISdk,
  session_id: string,
): Promise<Record<string, unknown>> {
  const v = await stateGet(iii, runRequestKey(session_id));
  return v && typeof v === 'object' ? (v as Record<string, unknown>) : {};
}

export async function saveCwd(iii: ISdk, session_id: string, cwd: string): Promise<void> {
  await stateSet(iii, cwdKey(session_id), cwd);
}

export async function saveCwdIndex(iii: ISdk, cwd_hash: string, session_id: string): Promise<void> {
  await stateSet(iii, cwdIndexKey(cwd_hash), session_id);
}

export async function loadSandboxId(iii: ISdk, session_id: string): Promise<string | null> {
  const v = await stateGet(iii, sandboxIdKey(session_id));
  return typeof v === 'string' ? v : null;
}

export async function saveFunctionSchemas(
  iii: ISdk,
  session_id: string,
  schemas: unknown,
): Promise<void> {
  await stateSet(iii, functionSchemasKey(session_id), schemas);
}

export async function loadFunctionSchemas(iii: ISdk, session_id: string): Promise<unknown[]> {
  const v = await stateGet(iii, functionSchemasKey(session_id));
  if (Array.isArray(v)) return v;
  const legacy = await stateGet(iii, toolSchemasKey(session_id));
  if (Array.isArray(legacy)) return legacy;
  return [];
}

const PREPARED_KEY = 'function_prepared';
const EXECUTED_KEY = 'function_executed';
const LEGACY_PREPARED_KEY = 'tool_prepared';
const LEGACY_EXECUTED_KEY = 'tool_executed';

const stagingKey = (sid: string, suffix: string) => `session/${sid}/${suffix}`;

async function stagingGetWithLegacy(
  iii: ISdk,
  session_id: string,
  newSuffix: string,
  legacySuffix: string,
): Promise<unknown[]> {
  const v = await stateGet(iii, stagingKey(session_id, newSuffix));
  if (Array.isArray(v)) return v;
  const legacy = await stateGet(iii, stagingKey(session_id, legacySuffix));
  return Array.isArray(legacy) ? legacy : [];
}

export type PreparedEntry = {
  function_call: FunctionCall;
  blocked: FunctionResult | null;
  pre_approved?: boolean;
};
export type ExecutedEntry = {
  function_call: FunctionCall;
  result: FunctionResult;
  is_error: boolean;
};

export async function savePreparedCalls(
  iii: ISdk,
  session_id: string,
  prepared: PreparedEntry[],
): Promise<void> {
  const payload = prepared.map((e) => ({
    function_call: e.function_call,
    blocked: e.blocked,
    pre_approved: e.pre_approved ?? false,
  }));
  await stateSet(iii, stagingKey(session_id, PREPARED_KEY), payload);
}

export async function loadPreparedCalls(iii: ISdk, session_id: string): Promise<PreparedEntry[]> {
  const items = await stagingGetWithLegacy(iii, session_id, PREPARED_KEY, LEGACY_PREPARED_KEY);
  const out: PreparedEntry[] = [];
  for (const it of items) {
    if (!it || typeof it !== 'object') continue;
    const obj = it as Record<string, unknown>;
    const fc = (obj.function_call ?? obj.tool_call) as FunctionCall | undefined;
    if (!fc) continue;
    const blocked = (obj.blocked as FunctionResult | null) ?? null;
    const pre_approved = obj.pre_approved === true;
    out.push({ function_call: fc, blocked, pre_approved });
  }
  return out;
}

export async function saveExecutedCalls(
  iii: ISdk,
  session_id: string,
  executed: ExecutedEntry[],
): Promise<void> {
  await stateSet(iii, stagingKey(session_id, EXECUTED_KEY), executed);
}

export async function loadExecutedCalls(iii: ISdk, session_id: string): Promise<ExecutedEntry[]> {
  const items = await stagingGetWithLegacy(iii, session_id, EXECUTED_KEY, LEGACY_EXECUTED_KEY);
  const out: ExecutedEntry[] = [];
  for (const it of items) {
    if (!it || typeof it !== 'object') continue;
    const obj = it as Record<string, unknown>;
    const fc = (obj.function_call ?? obj.tool_call) as FunctionCall | undefined;
    const result = obj.result as FunctionResult | undefined;
    if (!fc || !result) continue;
    out.push({
      function_call: fc,
      result,
      is_error: typeof obj.is_error === 'boolean' ? obj.is_error : false,
    });
  }
  return out;
}

export function findExecutedCall(
  executed: ExecutedEntry[],
  function_call_id: string,
): ExecutedEntry | undefined {
  return executed.find((e) => e.function_call.id === function_call_id);
}

export function upsertExecutedCall(executed: ExecutedEntry[], entry: ExecutedEntry): void {
  const idx = executed.findIndex((e) => e.function_call.id === entry.function_call.id);
  if (idx >= 0) executed[idx] = entry;
  else executed.push(entry);
}
